### PR TITLE
[TR] PIM-4781: Move User class from Oro to PIM

### PR DESCRIPTION
**Pull request by solivier**
| Q                  | A
| ------------------ | ---
| Migration script?  | n
| Behats             | y
| Specs & Static     | y
| Changelog updated  | n

- [x] rebase commits
- [x] remove commit for CI
- [x] Check for EE

http://core-ci.akeneo.com/job/pim-pr-behat/3152/console

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -51,7 +51,7 @@
         "monolog/monolog": "1.10.0",
         "nelmio/api-doc-bundle": "2.9.0",
         "oneup/flysystem-bundle": "1.1.0",
-        "oro/platform": "dev-pim-1.4",
+        "oro/platform": "dev-user-override",
```
Comment (solivier) [inline composer.json]: Don't forget to remove

File: src/Pim/Bundle/UserBundle/Entity/User.php [hunk 2]
```diff
@@ -0,0 +1,1041 @@
...
+    /** @var string */
+    protected $api;
+
+    /**  @var \DateTime $createdAt */
+    protected $createdAt;
+
+    /** @var \DateTime $updatedAt */
+    protected $updatedAt;
+
+    /** @var Pim\Bundle\CatalogBundle\Entity\Locale */
```
Comment (grena) [inline src/Pim/Bundle/UserBundle/Entity/User.php]: No FQCN for phpdoc please

File: src/Pim/Bundle/UserBundle/Entity/User.php [hunk 3]
```diff
@@ -0,0 +1,1041 @@
...
+    protected $catalogLocale;
+
+    /** @var Pim\Bundle\CatalogBundle\Entity\Channel */
```
Comment (grena) [inline src/Pim/Bundle/UserBundle/Entity/User.php]: No FQCN for phpdoc please

File: src/Pim/Bundle/UserBundle/Entity/User.php [hunk 4]
```diff
@@ -0,0 +1,1041 @@
...
+    protected $catalogScope;
+
+    /** @var Pim\Bundle\CatalogBundle\Model\CategoryInterface */
```
Comment (grena) [inline src/Pim/Bundle/UserBundle/Entity/User.php]: No FQCN for phpdoc please

File: src/Pim/Bundle/UserBundle/Entity/User.php [hunk 5]
```diff
@@ -0,0 +1,1041 @@
...
+    protected $defaultTree;
+
+    /**
+     * Constructor
```
Comment (grena) [inline src/Pim/Bundle/UserBundle/Entity/User.php]: delete it

File: src/Pim/Bundle/UserBundle/Entity/User.php [hunk 6]
```diff
@@ -0,0 +1,1041 @@
...
+    {
+        $this->nameSuffix = $nameSuffix;
+    }
+
+    /**
+     *
+     * @param  DateTime $birthday [optional] New birthday value. Null by default.
+     * @return User
+     */
+    public function setBirthday(DateTime $birthday = null)
```
Comment (grena) [inline src/Pim/Bundle/UserBundle/Entity/User.php]: :cake:

**Comment by grena** [comment]
GTM after comments fixed

File: features/Context/FixturesContext.php [hunk 7]
```diff
@@ -1446,7 +1446,7 @@ public function getProduct($sku)
     /**
      * @param string $username
      *
-     * @return \Oro\Bundle\UserBundle\Entity\User
+     * @return \Pim\Bundle\UserBundle\Entity\User
```
Comment (phaseinducer) [inline features/Context/FixturesContext.php]: Can you import please ?

File: src/Pim/Bundle/UserBundle/Entity/User.php [hunk 8]
```diff
@@ -0,0 +1,1041 @@
...
+    protected $defaultTree;
+
+    /**
+     * Constructor
+     */
+    public function __construct()
+    {
+        $this->salt = base_convert(sha1(uniqid(mt_rand(), true)), 16, 36);
+        $this->roles = new ArrayCollection();
+        $this->groups = new ArrayCollection();
```
Comment (skeleton) [inline src/Pim/Bundle/UserBundle/Entity/User.php]: please align

File: src/Pim/Bundle/UserBundle/Entity/User.php [hunk 9]
```diff
@@ -0,0 +1,1041 @@
...
+
+    /**
+     * Random string sent to the user email address in order to verify it
+     *
+     * @var string
+     */
+    protected $confirmationToken;
+
+    /**
+     * @var DateTime
```
Comment (phaseinducer) [inline src/Pim/Bundle/UserBundle/Entity/User.php]: single line

File: src/Pim/Bundle/UserBundle/Entity/User.php [hunk 10]
```diff
@@ -0,0 +1,1041 @@
...
+
+    /**
+     * @var Group[]
+     */
+    protected $groups;
+
+    /** @var string */
+    protected $api;
+
+    /**  @var \DateTime $createdAt */
```
Comment (phaseinducer) [inline src/Pim/Bundle/UserBundle/Entity/User.php]: backslash is useless if the class is imported

File: src/Pim/Bundle/UserBundle/Entity/User.php (hunk 8 above)
Comment (phaseinducer) [inline src/Pim/Bundle/UserBundle/Entity/User.php]: align please

File: src/Pim/Bundle/UserBundle/Entity/User.php [hunk 11]
```diff
@@ -0,0 +1,1041 @@
...
+    public function unserialize($serialized)
+    {
+        list(
+            $this->password,
+            $this->salt,
+            $this->username,
+            $this->enabled,
+            $this->confirmationToken,
+            $this->id
+            ) = unserialize($serialized);
```
Comment (phaseinducer) [inline src/Pim/Bundle/UserBundle/Entity/User.php]: fix alignement please

File: src/Pim/Bundle/UserBundle/Entity/User.php [hunk 12]
```diff
@@ -0,0 +1,1041 @@
...
+     */
+    public function getNameSuffix()
+    {
+        return $this->nameSuffix;
+    }
+
+    /**
+     * Return birthday
+     *
+     * @return \DateTime
```
Comment (phaseinducer) [inline src/Pim/Bundle/UserBundle/Entity/User.php]: no backslash, or remove import

File: src/Pim/Bundle/UserBundle/Entity/User.php [hunk 13]
```diff
@@ -0,0 +1,1041 @@
...
+
+    public function isPasswordRequestNonExpired($ttl)
+    {
+        return $this->getPasswordRequestedAt() instanceof DateTime &&
+               $this->getPasswordRequestedAt()->getTimestamp() + $ttl > time();
+    }
+
+    /**
+     * @param int $id
+     * @return mixed
```
Comment (phaseinducer) [inline src/Pim/Bundle/UserBundle/Entity/User.php]: blank line please, same for following methods

File: src/Pim/Bundle/UserBundle/Entity/User.php [hunk 14]
```diff
@@ -0,0 +1,1041 @@
...
+
+    /**
+     * {@inheritDoc}
+     */
+    public function isEnabled()
+    {
+        return $this->enabled;
+    }
+
+    public function isAccountNonExpired()
```
Comment (phaseinducer) [inline src/Pim/Bundle/UserBundle/Entity/User.php]: no phpdoc

File: src/Pim/Bundle/UserBundle/Entity/User.php [hunk 15]
```diff
@@ -0,0 +1,1041 @@
...
+    /**
+     * Pre persist event listener
+     *
+     * @ORM\PrePersist
+     */
+    public function beforeSave()
+    {
+        $this->createdAt = new DateTime('now', new \DateTimeZone('UTC'));
+        $this->updatedAt = new DateTime('now', new \DateTimeZone('UTC'));
+        $this->loginCount = 0;
```
Comment (phaseinducer) [inline src/Pim/Bundle/UserBundle/Entity/User.php]: align

**Pull request merged**